### PR TITLE
New StringParse* functions

### PR DESCRIPTION
**Pull request by inashivb**
Add a new set of StringParse* functions and use them for strict parsing in all parser, configuration and setup functions.
Related to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/3053

### CODE REVIEW

**Comment by victorjulien** [comment]
So the error checking by the updated callers caught my eye. This is not something you updated here, but it does look kind of funny. We see:
`if (StringParse...) <= 0)` -> error
`if ((StringParse...) == -1)` -> error
`if ((StringParse...) < 0)` -> error
`if ((StringParse...) != strlen(..))` -> error
`if ((StringParse...) > 0)` -> no error

Header says:
```
* \return n Number of bytes extracted on success
* \return -1 On error
```
So what is the correct pattern?

**Comment by inashivb** [comment]
> So the error checking by the updated callers caught my eye. This is not something you updated here, but it does look kind of funny. We see:
> `if (StringParse...) <= 0)` -> error
> `if ((StringParse...) == -1)` -> error
> `if ((StringParse...) < 0)` -> error
> `if ((StringParse...) != strlen(..))` -> error
> `if ((StringParse...) > 0)` -> no error
> 
> Header says:
> 
> ```
> * \return n Number of bytes extracted on success
> * \return -1 On error
> ```
> 
> So what is the correct pattern?

Seems like Error should be whenever the retval is -1.
I do  not know about `if ((StringParse...) != strlen(..))`. At some places, the statements are like
`if (StringParseUint16(rep_value, 10, 0, key) != (int)strlen(key))`
not sure what this means. That if key = "2", then it should fail because the length of key is 1?

**Comment by victorjulien** [comment]
> > So the error checking by the updated callers caught my eye. This is not something you updated here, but it does look kind of funny. We see:
> > `if (StringParse...) <= 0)` -> error
> > `if ((StringParse...) == -1)` -> error
> > `if ((StringParse...) < 0)` -> error
> > `if ((StringParse...) != strlen(..))` -> error
> > `if ((StringParse...) > 0)` -> no error
> > Header says:
> > ```
> > * \return n Number of bytes extracted on success
> > * \return -1 On error
> > ```
> > 
> > 
> > So what is the correct pattern?
> 
> Seems like Error should be whenever the retval is -1.
> I do not know about `if ((StringParse...) != strlen(..))`. At some places, the statements are like
> `if (StringParseUint16(rep_value, 10, 0, key) != (int)strlen(key))`
> not sure what this means. That if key = "2", then it should fail because the length of key is 1?

i think it means that if the function returns the number of characters it consumed in case of success, we only consider it success if it consumed **all** the characters we fed it.

File: src/util-byte.c [hunk 1]
```diff
@@ -223,29 +223,25 @@ int ByteExtractString(uint64_t *res, int base, uint16_t len, const char *str)
...
     }
-    /* This will interfere with some rules that do not know the length
-     * in advance and instead are just using the max.
-     */
-#if 0
-    else if (len && *endptr != '\0') {
-        fprintf(stderr, "ByteExtractString: Extra characters following numeric value\n");
+    else if (strict && len && *endptr != '\0') {
+        SCLogError(SC_ERR_INVALID_NUMERIC_VALUE, "ByteExtractString: "
+                   "Extra characters following numeric value\n");
```
Comment (victorjulien) [inline src/util-byte.c]: Please remove the function name and the trailing \n here

File: src/util-byte.c [hunk 2]
```diff
@@ -337,15 +397,11 @@ int ByteExtractStringSigned(int64_t *res, int base, uint16_t len, const char *st
...
         return -1;
     }
-    /* This will interfere with some rules that do not know the length
-     * in advance and instead are just using the max.
-     */
-#if 0
-    else if (len && *endptr != '\0') {
-        fprintf(stderr, "ByteExtractStringSigned: Extra characters following numeric value\n");
+    else if (strict && len && *endptr != '\0') {
+        SCLogError(SC_ERR_INVALID_NUMERIC_VALUE, "ByteExtractStringSigned: "
```
Comment (victorjulien) [inline src/util-byte.c]: remove func name and \n

File: src/util-byte.c [hunk 3]
```diff
@@ -420,6 +476,73 @@ int ByteExtractStringInt8(int8_t *res, int base, uint16_t len, const char *str)
...
+
+    ret = ByteExtractStringSigned(&i64, base, len, str, true);
+    if (ret <= 0) {
+        return ret;
+    }
+
+    *res = (int32_t)i64;
+
+    if ((int64_t)(*res) != i64) {
+        SCLogError(SC_ERR_NUMERIC_VALUE_ERANGE, "Numeric value out of range "
```
Comment (victorjulien) [inline src/util-byte.c]: why are these `SCLogError` here but `SCLogDebug` for the unsigned function variants?

**Comment by victorjulien** [comment]
> > Seems like Error should be whenever the retval is -1.
> > I do not know about `if ((StringParse...) != strlen(..))`. At some places, the statements are like
> > `if (StringParseUint16(rep_value, 10, 0, key) != (int)strlen(key))`
> > not sure what this means. That if key = "2", then it should fail because the length of key is 1?
> 
> i think it means that if the function returns the number of characters it consumed in case of success, we only consider it success if it consumed **all** the characters we fed it.

I suppose in strict mode this is implied. But in non-strict mode it is a way for the caller to know if there is trailing data.

File: src/util-byte.c (hunk 3 above)
Comment (inashivb) [reply src/util-byte.c]: Copy paste issue. Fixed now.

**Comment by inashivb** [comment]
Replaced with #4646

**Pull request closed without merging**